### PR TITLE
Optimize paste hot path and add edge-case tests

### PR DESCRIPTION
**Pull request by appunni-m**
• Here’s what changed in paste:

  Core implementation changes

  - In src/utils.rs:105, paste_with_mask now has typed fast paths for common modes:
  - Mask handling now reuses existing grayscale buffers when possible to reduce allocations: src/utils.rs:124.
  - Blending switched from float math to integer math via blend_u8 for lower overhead and stable rounding: src/utils.rs:234.
  - Added alpha short-circuiting in fast paths:
      - alpha == 0 skip pixel
      - alpha == 255 direct copy
        (see src/utils.rs:252, src/utils.rs:284)
  - Kept generic fallback behavior for uncommon mode combinations so compatibility is preserved: src/utils.rs:137.

  Color fill path changes

  - fill_region now also uses typed fast paths (RGB, RGBA, L) before fallback, reducing per-pixel dynamic overhead: src/utils.rs:173, src/utils.rs:188.

  New paste edge-case tests
  Added 3 tests in python/puhu/tests/test_paste.py:193:

  - Mask size mismatch should raise.
  - Color paste with only 2-tuple box should raise.
  - Fully transparent mask should be a no-op.

  Validation run: 15 passed for test_paste.py.

### CODE REVIEW

**Comment by appunni-m** [comment]
I re verified it against pillow to be Safe I am not breaking anything. May I suggest and integration test suite which validates certain version of Pillow is compatible with current version of Puhu, with backward compatibility to N different Pillow Version. 

Also does Puhu also acts as a Rust library ?

File: src/image.rs [hunk 1]
```diff
@@ -554,9 +554,7 @@ impl PyImage {
             PasteSource::Image(src_img) => {
                 // Mode conversion if needed
                 let src_mode = color_type_to_mode_string(src_img.color());
-                let source_converted = if dest_mode != src_mode
-                    && !(dest_mode == "RGB" && matches!(src_mode.as_str(), "RGBA" | "LA" | "RGBa"))
-                {
+                let source_converted = if dest_mode != src_mode {
```
Comment (bgunebakan) [inline src/image.rs:557]: Good catch! I added this check to align with pillow but I double checked the **image** crate and **overlay()** naturally respects alpha channels and performs alpha blending, so we don't need this here.

File: src/utils.rs [hunk 2]
```diff
@@ -121,35 +124,41 @@ pub fn paste_with_mask(
         None => return Ok(()), // Nothing to paste
     };
 
-    // Convert mask to grayscale if needed
-    let mask_gray = mask.to_luma8();
+    // Reuse grayscale masks without reallocation when possible.
+    let mask_gray = mask.as_luma8().cloned().unwrap_or_else(|| mask.to_luma8());
```
Comment (bgunebakan) [inline src/utils.rs:128]: The intention was clearly to avoid reallocations when the mask is already a Luma8 layout. Calling **.cloned()** forces a deep copy of the image buffer each time, that defeating the purpose. This will cause performance issues.

File: src/utils.rs [hunk 3]
```diff
@@ -178,19 +187,131 @@ pub fn fill_region(
...
+                continue;
+            }
+            if alpha == 255 {
+                *dest.get_pixel_mut(dx, dy) = *src.get_pixel(sx, sy);
+                continue;
+            }
+            let inv_alpha = 255u16 - alpha as u16;
+            let s = src.get_pixel(sx, sy).0;
+            let d = dest.get_pixel(dx, dy).0;
+            *dest.get_pixel_mut(dx, dy) = image::Rgb([
```
Comment (bgunebakan) [inline src/utils.rs:273]: **get_pixel_mut(dx, dy)** is called right after on the same coordinates. This causes to evaluate the buffer bounds twice per pixel iteration. my suggestion would be to grab the mutable pixel reference upfront, then capture its inner value to compute the blend.
```
let s = src.get_pixel(sx, sy).0;
let d_pixel = dest.get_pixel_mut(dx, dy);
let d = d_pixel.0;

*d_pixel = image::Rgb([
    blend_u8(s[0], d[0], alpha, inv_alpha),
    blend_u8(s[1], d[1], alpha, inv_alpha),
    blend_u8(s[2], d[2], alpha, inv_alpha),
]);

```

**Review by bgunebakan — CHANGES_REQUESTED** [review]
Thank you so much for your contribution! I added some comments about some small points. Best!

**Comment by bgunebakan** [comment]
> I re verified it against pillow to be Safe I am not breaking anything. May I suggest and integration test suite which validates certain version of Pillow is compatible with current version of Puhu, with backward compatibility to N different Pillow Version.
> 
> Also does Puhu also acts as a Rust library ?

Thanks! I was thinking to implement same on pipeline so we can see compability on every new feature implementation. I have a small script that shows compability but does not cover all methods. I compare to pillow 12 docs and codebase. What do you have in your mind?

**Comment by appunni-m** [comment]
> Thanks! I was thinking to implement same on pipeline so we can see compability on every new feature implementation. I have a small script that shows compability but does not cover all methods. I compare to pillow 12 docs and codebase. What do you have in your mind?

I was thinking if we have a dev testing setup where same test case tests pillow or your library (inject library via monkey patch) for each set of test it represent one capability. Also have a way to publish it's outcome to a doc so that it's a living doc, comparing multiple pillow versions against each other and the library. I can send a new PR if you are interested

**Comment by bgunebakan** [comment]
Hi @appunni-m, I'd like to merge this PR and close the issue. We can keep track of other features in other PR's. Would you like to push changes for my comments?

Thanks.
